### PR TITLE
Fix RT-DETR exported onnx model

### PR DESCRIPTION
**Pull request by Kayzwer**
fix RT-DETR exported onnx model, cast data in the where operations to float32

### Summary
🐛🔧🚀

Fix type mismatch errors in `Head` module when using mixed precision training. Ensure tensors with constants have `torch.float32` dtype in `ultralytics/nn/modules/head.py`.

> _`torch.where` fails_
> _with mixed precision training_
> _snowflakes of `float32`_

### Walkthrough
*  Fix type mismatch errors when using torch.where with torch.inf or 0 in `head.py` ([link](https://github.com/ultralytics/ultralytics/pull/3317/files?diff=unified&w=0#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639L273-R273), [link](https://github.com/ultralytics/ultralytics/pull/3317/files?diff=unified&w=0#diff-06038a7c9e9c565760881118b974c2f2fe7aee7c78bb6c142d1498d0d9614639L297-R297))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved anchor generation and feature encoding in Ultralytics neural network module. 

### 📊 Key Changes
- Replaced `torch.where` with `masked_fill` for anchor adjustment.
- Simplified feature masking using element-wise multiplication instead of `torch.where`.

### 🎯 Purpose & Impact
- 🚀 Improves code clarity and potentially enhances computational efficiency.
- 🛠️ These changes ensure better maintenance and readability of the code, which is crucial for developers working on the project.
- 🔄 Users might experience subtle performance improvements due to more efficient operations without any direct changes to functionality.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@Kayzwer thanks for the PR! Will this support `yolo export format=onnx half=True` on GPU?

**Comment by Kayzwer** [comment]
for half, I tried on yolov8 but does not work properly on .net onnxruntime

**Comment by glenn-jocher** [comment]
@Kayzwer ok I made some simplifications. This works now for FP32 exports but I still get an error when simplifying FP16 exports. Do you know where this new Range error is coming from?
```bash
yolo export format=onnx device=0 model=rtdetr-l.pt simplify half
```

Screenshot 2023-06-22 at 18 11 31
